### PR TITLE
[danger] Add a check for changes that affect multiple components.

### PR DESCRIPTION
**Pull request by jverkoey**
Adds a new check to our Danger CI that provides a warning if a PR is affecting the source code of multiple components simultaneously.

### CODE REVIEW

**Comment by romoore** [comment]
Please add a description.

**Comment by jverkoey** [comment]
WIP!

**Comment by romoore** [comment]
Sorry, I got an email asking for review.

**Comment by jverkoey** [comment]
Ready for review. Will revert the source modifications once the danger CI passes.

File: dangerfile.js [hunk 1]
```diff
@@ -18,9 +18,46 @@ const {danger, warn, fail} = require('danger')
...
+  while (path_parts[0][0] == path_parts[0][0].toLowerCase()) {
+    component_path.push(path_parts.splice(0, 1));
+  }
+  // Only return a string if we actually found a component folder.
+  if (path_parts[0][0] == path_parts[0][0].toUpperCase()) {
+    component_path.push(path_parts.splice(0, 1));
+    return component_path.join('/');
+  }
+  return null;
+}).filter(function(path) { return path !== null; })));
```
Comment (romoore) [inline dangerfile.js:61]: null.filter? I'm not a JS expert, but is that ok?

File: dangerfile.js [hunk 2]
```diff
@@ -18,9 +18,46 @@ const {danger, warn, fail} = require('danger')
...
+  return path.startsWith('components/') && path.indexOf('/src') >= 0;
+});
+
+// Generate a list of unique components that have been modified.
+let components = Array.from(new Set(component_files.map(function(path) {
+  var path_parts = path.split('/');
+  path_parts.splice(0, 1); // Drop the components/ prefix.
+  
+  var component_path = [];
+  // Convention: lower-case names are directories. Upper-case names are components.
```
Comment (romoore) [inline dangerfile.js]: Where is this convention used? In the source, we use lowercase and Camel Case, so it's `components/ActivityIndicator`, not `components/ACTIVITYINDICATOR`.

File: dangerfile.js (hunk 1 above)
Comment (jverkoey) [reply dangerfile.js:61]: The return value from function above is a list of paths or nulls. The filter removes the nulls (this is essentially `flatMap` in swift).

File: dangerfile.js (hunk 2 above)
Comment (jverkoey) [reply dangerfile.js]: Updated the docs to more clearly document the convention.

File: dangerfile.js [hunk 3]
```diff
@@ -18,9 +18,49 @@ const {danger, warn, fail} = require('danger')
...
+// Generate a list of unique components that have been modified.
+let components = Array.from(new Set(component_files.map(function(path) {
+  var path_parts = path.split('/');
+  path_parts.splice(0, 1); // Drop the components/ prefix.
+  
+  var component_path = [];
+  // Convention: all-lower-case names are directories, e.g. `components/private/`.
+  // Capitalized names are components, e.g. `BottomNavigation`.
+  // We want to get the full path up to and including the first component folder.
+  while (path_parts[0][0] == path_parts[0][0].toLowerCase()) {
```
Comment (romoore) [inline dangerfile.js:50]: If we're checking for "all lowercase", should we lowercase the string and check that instead of only the first character?

**Review by romoore — APPROVED** [review]

File: dangerfile.js (hunk 3 above)
Comment (jverkoey) [reply dangerfile.js:50]: We only care if the first character is lowercase — the convention is "lower-case first letter is a folder", "upper-case first letter is a component".

Comment (romoore) [reply dangerfile.js:50]: That's what I thought, but the comment referred to "all-lower-case names".

**Pull request merged**